File: src/main/java/fr/adrienbrault/idea/symfony2plugin/routing/inspection/PhpRouteMissingInspection.java
package fr.adrienbrault.idea.symfony2plugin.routing.inspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.patterns.ElementPattern;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.routing.PhpRouteReferenceContributor;
import fr.adrienbrault.idea.symfony2plugin.routing.RouteHelper;
import fr.adrienbrault.idea.symfony2plugin.util.MethodMatcher;
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class PhpRouteMissingInspection extends LocalInspectionTool {
    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if(!Symfony2ProjectComponent.isEnabled(holder.getProject())) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        return new MyPsiElementVisitor(holder);
    }

    private static class MyPsiElementVisitor extends PsiElementVisitor {
        @NotNull private final ProblemsHolder holder;
        private ElementPattern<?> methodWithFirstStringPattern;

        MyPsiElementVisitor(@NotNull ProblemsHolder holder) {
            this.holder = holder;
        }

        @Override
        public void visitElement(@NotNull PsiElement element) {
            if(getMethodWithFirstStringPattern().accepts(element)) {
                String contents = PhpElementsUtil.getStringValue(element);
                if(StringUtils.isNotBlank(contents)) {
                    invoke(contents, element, holder);
                }
            }

            super.visitElement(element);
        }

        private ElementPattern<?> getMethodWithFirstStringPattern() {
            return methodWithFirstStringPattern != null ? methodWithFirstStringPattern : (methodWithFirstStringPattern = PhpElementsUtil.getMethodWithFirstStringOrNamedArgumentPattern());
        }
    }

    private static void invoke(@NotNull String routeName, @NotNull final PsiElement element, @NotNull ProblemsHolder holder) {
        MethodMatcher.MethodMatchParameter methodMatchParameter = new MethodMatcher.StringParameterMatcher(element, 0)
            .withSignature(PhpRouteReferenceContributor.GENERATOR_SIGNATURES)
            .match();

        if(methodMatchParameter == null) {
            return;
        }

        if (!RouteHelper.isExistingRouteName(element.getProject(), routeName)) {
            holder.registerProblem(element, "Symfony: Missing Route", ProblemHighlightType.GENERIC_ERROR_OR_WARNING, new RouteGuessTypoQuickFix(routeName));
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/routing/inspection/PhpRouteMissingInspection.kt
package fr.adrienbrault.idea.symfony2plugin.routing.inspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.patterns.ElementPattern
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.routing.PhpRouteReferenceContributor
import fr.adrienbrault.idea.symfony2plugin.routing.RouteHelper
import fr.adrienbrault.idea.symfony2plugin.util.MethodMatcher
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil
import org.apache.commons.lang3.StringUtils

private fun inspectPhpRouteMissing(routeName: String, element: PsiElement, holder: ProblemsHolder) {
    val methodMatchParameter: MethodMatcher.MethodMatchParameter? = MethodMatcher.StringParameterMatcher(element, 0)
        .withSignature(PhpRouteReferenceContributor.GENERATOR_SIGNATURES)
        .match()

    if (methodMatchParameter == null) {
        return
    }

    if (!RouteHelper.isExistingRouteName(element.project, routeName)) {
        holder.registerProblem(element, "Symfony: Missing Route", ProblemHighlightType.GENERIC_ERROR_OR_WARNING, RouteGuessTypoQuickFix(routeName))
        return
    }

    if (RouteHelper.isRouteControllerDeprecated(element.project, routeName)) {
        holder.registerProblem(element, "Symfony: Controller action is deprecated", ProblemHighlightType.LIKE_DEPRECATED)
    }
}

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class PhpRouteMissingInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return MyPsiElementVisitor(holder)
    }

    private class MyPsiElementVisitor(private val holder: ProblemsHolder) : PsiElementVisitor() {
        private var methodWithFirstStringPattern: ElementPattern<*>? = null

        override fun visitElement(element: PsiElement) {
            if (element is StringLiteralExpression && getMethodWithFirstStringPattern().accepts(element)) {
                val contents = PhpElementsUtil.getStringValue(element)
                if (StringUtils.isNotBlank(contents)) {
                    inspectPhpRouteMissing(contents!!, element, holder)
                }
            }

            super.visitElement(element)
        }

        private fun getMethodWithFirstStringPattern(): ElementPattern<*> {
            if (methodWithFirstStringPattern != null) {
                return methodWithFirstStringPattern!!
            }

            methodWithFirstStringPattern = PhpElementsUtil.getMethodWithFirstStringOrNamedArgumentPattern()
            return methodWithFirstStringPattern!!
        }
    }
}
